### ISSUE TITLE
ReceiveAllAsync, ReadAllAsync and propagation of errors

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#79535]
Hi! Recently I tried to use the [`DataflowBlock.ReceiveAllAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.dataflowblock.receiveallasync) API for the first time, but I ended up not using it because it didn't have the desirable behavior [for my case](https://stackoverflow.com/questions/74756252/custom-dataflow-transform-block-with-dynamic-maxdegreeofparallelism "Custom Dataflow transform block with dynamic MaxDegreeOfParallelism"). My expectation was that after completing the enumeration of the asynchronous sequence, an exception stored in the consumed `BufferBlock<T>` would be surfaced. But this is not what happens. Here is a minimal demo:

```C#
BufferBlock<int> block = new();
((IDataflowBlock)block).Fault(new Exception());
try
{
    await foreach (var item in block.ReceiveAllAsync()) { }
    Console.WriteLine($"Done: {block.Completion.Status}");
}
catch (Exception ex) { Console.WriteLine(ex.Message); }
```

Actual output:

```
Done: Faulted
```

The output that I expected was:

```
Exception of type 'System.Exception' was thrown.
```

The `ReceiveAllAsync` API was introduced in .NET 6. My expectation for its behavior was based on the preexisting API [`ChannelReader<T>.ReadAllAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.channels.channelreader-1.readallasync) (Core 3.0+), which plays a similar role in the Channels library. That API is used to consume a `Channel<T>`, and it **does** propagate the error that is stored in the channel. Here is a minimal demo:

```C#
Channel<int> channel = Channel.CreateUnbounded<int>();
channel.Writer.Complete(new Exception());
try
{
    await foreach (var item in channel.Reader.ReadAllAsync()) { }
    Console.WriteLine($"Done: {channel.Reader.Completion.Status}");
}
catch (Exception ex) { Console.WriteLine(ex.Message); }
```

Output:

```
Exception of type 'System.Exception' was thrown.
```

[Both demos online](https://dotnetfiddle.net/VnJaOA).

My perception is that the [`Channel<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.channels.channel-1) class is an optimized version of the older [`BufferBlock<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.dataflow.bufferblock-1) class, with some extra features (like the ability to drop old messages when the channel is full). So I was surprised by the difference in behavior. I would like to ask if the non-error-propagating behavior of the newer API (`ReceiveAllAsync`) is by design. Also, in case it is, I would be interested to know whether the behavioral difference reflects a difference on how the two components (`BufferBlock<T>` & `Channel<T>`) are intended to be used in producer-consumer scenarios.

Thanks!

**Comment by theodorzoulias** [comment]
Btw I think that the correct label for this issue is **[area-System.Threading.Tasks](https://github.com/dotnet/runtime/issues?q=sort%3Aupdated-desc+label%3Aarea-System.Threading.Tasks)**.

**Comment by stephentoub** [comment]
> So I was surprised by the difference in behavior. I would like to ask if the non-error-propagating behavior of the newer API (ReceiveAllAsync) is by design.

Yes and no.  It's by design only in that it's inheriting the behavior of OutputAvailableAsync.  Whereas ChannelReader's WaitToReadAsync will fail with an exception if the channel has faulted, OutputAvailableAsync doesn't propagate exceptions from the block, so if the block has completed with a failure, the task returned from OutputAvailableAsync just yields false rather than faults.

We could consider taking a breaking change to await the block's Completion task upon exiting the loop to accomodate the difference.  I don't have a good sense for how many folks we'd break, as the method was shipped new in the .NET 6 timeframe.  It's also simple enough for the consumer to do exactly the same thing if they care about the exception being propagated.

**Comment by theodorzoulias** [comment]
> I don't have a good sense for how many folks we'd break, as the method was shipped new in the .NET 6 timeframe.

In StackOverflow [there are currently three posts](https://stackoverflow.com/search?q=%5Bc%23%5D+ReceiveAllAsync "Results for receiveallasync tagged with c#") that are mentioning the `ReceiveAllAsync` method, one by @pkanavos and two by myself. My guess is that few people have used it, even fewer have used it with blocks that might fail, and even fewer will have their code broken if this bug is fixed. It's not impossible that by fixing this bug, some currently buggy code will be fixed.
